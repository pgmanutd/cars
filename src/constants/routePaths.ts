import { StockNumberParam } from 'features/CarInfo/types';

const routePaths = {
  cars: '/',
  carDetails: {
    pathWithParams: ({ stockNumber }: { stockNumber: StockNumberParam }) =>
      `/details/${stockNumber}`,
    path: '/details/:stockNumber',
  },
  favorites: '/favorites',
  notFound: '/not-found',
};

export default routePaths;
