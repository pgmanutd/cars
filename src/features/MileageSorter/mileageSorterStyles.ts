import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    formControl: {
      width: '100%',
      marginBottom: theme.spacing(3),
    },
    selectEmpty: {
      marginTop: theme.spacing(3),
    },
  }),
);
