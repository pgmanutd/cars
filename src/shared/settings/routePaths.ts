const routePaths = {
  cars: '/',
};

export default routePaths;
