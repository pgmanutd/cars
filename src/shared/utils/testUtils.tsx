import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
