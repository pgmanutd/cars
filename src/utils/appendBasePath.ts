const appendBasePath = (basePath: string, path: string) =>
  `${basePath}?${path}`;

export default appendBasePath;
