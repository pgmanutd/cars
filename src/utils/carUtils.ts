import { Car } from 'features/CarInfo/types';

export const getCarName = ({ manufacturerName, modelName }: Car) =>
  `${manufacturerName} ${modelName}`;

export const getCarFeatures = (
  { stockNumber, mileage, fuelType, color }: Car,
  { stockNumberPrefix = '' }: { stockNumberPrefix?: string } = {
    stockNumberPrefix: '',
  },
) =>
  `${stockNumberPrefix}${stockNumber} - ${mileage?.number} ${mileage?.unit} - ${fuelType} - ${color}`;
