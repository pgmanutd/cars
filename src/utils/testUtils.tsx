import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
