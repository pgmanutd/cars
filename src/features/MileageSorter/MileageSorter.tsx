import React, { memo, useCallback } from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import { useHistory } from 'react-router-dom';

import routePaths from 'constants/routePaths';
import { QUERY_KEYS } from 'constants/appConstants';

import appendBasePath from 'utils/appendBasePath';
import pickTargetValue from 'utils/pickTargetValue';

import useSetURLQuery from 'hooks/useSetURLQuery';

import { useTranslate } from 'features/Translate';

import { SortQuery } from './types';
import { useStyles } from './mileageSorterStyles';

export interface MileageSorterProps {
  sort: SortQuery;
}

const MileageSorter: React.FC<MileageSorterProps> = ({
  sort,
  ...restProps
}) => {
  const classes = useStyles();
  const { translate } = useTranslate();
  const history = useHistory();

  const sortParams = useSetURLQuery({
    [QUERY_KEYS.sort]: sort,
  });

  const handleSortChange = useCallback(
    (event: React.ChangeEvent<{ value: unknown }>) => {
      const updatedSortParams = new URLSearchParams(sortParams);

      updatedSortParams.set(QUERY_KEYS.sort, pickTargetValue(event) as string);

      history.push(
        appendBasePath(routePaths.cars, updatedSortParams?.toString()),
      );
    },
    [history, sortParams],
  );

  return (
    <FormControl
      data-testid="MileageSorter"
      {...restProps}
      className={classes.formControl}
    >
      <InputLabel shrink id="MileageSorterLabel">
        {translate('features.MileageSorter.sortLabel')}
      </InputLabel>
      <Select
        labelId="MileageSorterLabel"
        id="MileageSorterLabelSelect"
        value={sort}
        onChange={handleSortChange}
        displayEmpty
        className={classes.selectEmpty}
      >
        <MenuItem value="">
          <em>{translate('features.MileageSorter.sortDefaultValue')}</em>
        </MenuItem>
        <MenuItem value="asc">
          <em>{translate('features.MileageSorter.sortMileageAsc')}</em>
        </MenuItem>
        <MenuItem value="des">
          <em>{translate('features.MileageSorter.sortMileageDesc')}</em>
        </MenuItem>
      </Select>
    </FormControl>
  );
};

export default memo(MileageSorter);
