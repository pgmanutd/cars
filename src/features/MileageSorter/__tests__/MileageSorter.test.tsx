import React from 'react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';

import routePaths from 'constants/routePaths';

import { renderWithProviders } from 'utils/testUtils';

import MileageSorter from '../MileageSorter';

describe('<MileageSorter />', () => {
  const setup = () => {
    const initialRoute = routePaths.cars;
    const sort = '';

    const { renderResult, history } = renderWithProviders(
      <Route path={initialRoute}>
        <MileageSorter sort={sort} />
      </Route>,
      {
        routerConfig: {
          route: initialRoute,
        },
      },
    );

    return {
      sort,
      renderResult,
      history,
    };
  };

  it('should render the component and matches it against stored snapshot', () => {
    const { renderResult } = setup();

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should add "sort" query to location path when "asc" item in dropdown is clicked', () => {
    const { renderResult, history } = setup();

    userEvent.click(renderResult.getByText('None'));
    userEvent.click(renderResult.getByText('Mileage - Ascending'));

    const searchParams = new URLSearchParams(history.location.search);

    expect(searchParams.toString()).toBe('sort=asc');
  });
});
