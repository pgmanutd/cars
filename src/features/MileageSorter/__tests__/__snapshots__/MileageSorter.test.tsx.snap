// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MileageSorter /> should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root makeStyles-formControl-1"
    data-testid="MileageSorter"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink="true"
      id="MileageSorterLabel"
    >
      Sort by
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectEmpty-2 MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="MileageSorterLabel MileageSorterLabelSelect"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="MileageSorterLabelSelect"
        role="button"
        tabindex="0"
      >
        <em>
          None
        </em>
      </div>
      <input
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
